### PR TITLE
feat: allow custom Dats client

### LINKED ISSUE #35: Allow to pass custom client instance

### PR DESCRIPTION
**Pull request by antoniomuso**
close #35

### CODE REVIEW

**Review by simonecorsi — CHANGES_REQUESTED** [review]
Implementation LGTM! 

Let's add test cases of the throws so we don't lose coverage!

**Comment by antoniomuso** [comment]
Miss types and Docs

File: index.js [hunk 1]
```diff
@@ -4,14 +4,15 @@ const fp = require('fastify-plugin');
 const { default: Client } = require('@immobiliarelabs/dats');
 const doc = require('@dnlup/doc');
 const { hrtime2ns, hrtime2ms, hrtime2s } = require('@dnlup/hrtime-utils');
+const STATSD_METHODS = ['counter', 'timing', 'gauge', 'set'];
```
Comment (JellyBellyDev) [inline index.js:7]: we would remove `on`?

Comment (antoniomuso) [reply index.js]: Yes, it is not a method of Dats, I don't know why it was there.

File: index.js [hunk 2]
```diff
@@ -101,17 +110,32 @@ module.exports = fp(
...
-                  bufferSize,
-                  bufferFlushTimeout,
-                  udpDnsCache,
-                  udpDnsCacheTTL,
-                  onError: onError,
-              })
-            : clientMock();
+        let stats;
+
+        if (customDatsClient) {
```
Comment (dnlup) [inline index.js:115]: I am afraid that with this logic we lose the setup of the mocked client when we pass the custom instance.

File: README.md [hunk 3]
```diff
@@ -138,8 +138,9 @@ This module exports a [plugin registration function](https://github.com/fastify/
 -   `bufferFlushTimeout`: Number. Metrics buffer flush timeout. See [dats](https://github.com/immobiliare/dats#new-clientoptions).
 -   `sampleInterval`: Number. Optional. Sample interval in `ms` used to gather process stats. Defaults to `1000`.
 -   `onError`: Function: `(err) => void`. Optional. This function to handle possible Dats errors. See [dats](https://github.com/immobiliare/dats#new-clientoptions). Default: `(err) => log(err)`
--   `udpDnsCache`: Boolean. Optional. Activate udpDnsCache. Default `true`.
--   `udpDnsCacheTTL`: Number. Optional. DNS cache Time to live of an entry in seconds. Default `120`.
+-   `udpDnsCache`: Boolean. Optional. Activate udpDnsCache. See [dats](https://github.com/immobiliare/dats#new-clientoptions). Default `true`.
+-   `udpDnsCacheTTL`: Number. Optional. See [dats](https://github.com/immobiliare/dats#new-clientoptions). DNS cache Time to live of an entry in seconds. Default `120`.
+-   `customDatsClient`: Dats instance. Optional. The custom statsd client, It must have the same interface of [dats](https://github.com/immobiliare/dats#new-clientoptions). If set, all dats parameters will not work.
```
Comment (dnlup) [inline README.md:143]: Are you referring to the dns parameters and similar? If so, I think it is implied and I would omit it. The point of the feature is to allow a custom dats instance, and by dats I mean our client, not a custom StatsD client. At least for  now 😉 . I would rephrase this as:

"A custom dats instance" and point to the documentation of dats as you have done already.

**Review by dnlup — CHANGES_REQUESTED** [review]
Great work, thanks. I left a couple of comments.

File: index.js (hunk 2 above)
Comment (dnlup) [reply index.js:115]: Never mind, there is no need to initialize a mock client when a custom one is passed.

File: README.md (hunk 3 above)
Comment (antoniomuso) [reply README.md:143]: Ok, I fix it.
